azure/scope/machinepoolmachine.go

```diff
@@ -261,6 +274,152 @@ func (s *MachinePoolMachineScope) UpdateStatus(ctx context.Context) error {
+		Namespace: s.AzureMachinePoolMachine.Namespace,
+	})
+
+	if err != nil {
+		s.Error(err, "Error creating a remote client while deleting Machine, won't retry")
+		return nil
+	}
+
+	kubeClient, err := kubernetes.NewForConfig(restConfig)
+	if err != nil {
+		s.Error(err, "Error creating a remote client while deleting Machine, won't retry")
+		return nil
+	}
+
+	drainer := &drain.Helper{
+		Client:              kubeClient,
+		Force:               true,
+		IgnoreAllDaemonSets: true,
+		DeleteLocalData:     true,
+		GracePeriodSeconds:  -1,
+		// If a pod is not evicted in 20 seconds, retry the eviction next time the
+		// machine gets reconciled again (to allow other machines to be reconciled).
+		Timeout: 20 * time.Second,
+		OnPodDeletedOrEvicted: func(pod *corev1.Pod, usingEviction bool) {
+			verbStr := "Deleted"
+			if usingEviction {
+				verbStr = "Evicted"
+			}
+			s.V(4).Info(fmt.Sprintf("%s pod from Node", verbStr),
+				"pod", fmt.Sprintf("%s/%s", pod.Name, pod.Namespace))
+		},
+		Out:    writer{klog.Info},
+		ErrOut: writer{klog.Error},
+		DryRun: false,
+	}
+
+	if noderefutil.IsNodeUnreachable(node) {
+		// When the node is unreachable and some pods are not evicted for as long as this timeout, we ignore them.
+		drainer.SkipWaitForDeleteTimeoutSeconds = 60 * 5 // 5 minutes
+	}
+
+	if err := drain.RunCordonOrUncordon(ctx, drainer, node, true); err != nil {
+		// Machine will be re-reconciled after a cordon failure.
+		return azure.WithTransientError(errors.Errorf("unable to cordon node %s: %v", node.Name, err), 20*time.Second)
+	}
+
+	if err := drain.RunNodeDrain(ctx, drainer, node.Name); err != nil {
+		// Machine will be re-reconciled after a drain failure.
+		return azure.WithTransientError(errors.Wrap(err, "Drain failed, retry in 20s"), 20*time.Second)
+	}
+
+	s.V(4).Info("Drain successful")
+	return nil
+}
+
+// isNodeDrainAllowed checks to see the node is excluded from draining or if the NodeDrainTimeout has expired.
+func (s *MachinePoolMachineScope) isNodeDrainAllowed() bool {
+	if _, exists := s.AzureMachinePoolMachine.ObjectMeta.Annotations[clusterv1.ExcludeNodeDrainingAnnotation]; exists {
+		return false
+	}
+
+	if s.nodeDrainTimeoutExceeded() {
+		return false
+	}
+
+	return true
+}
+
+// nodeDrainTimeoutExceeded will check to see if the AzureMachinePool's NodeDrainTimeout is exceeded for the
+// AzureMachinePoolMachine.
+func (s *MachinePoolMachineScope) nodeDrainTimeoutExceeded() bool {
+	// if the NodeDrainTineout type is not set by user
+	pool := s.AzureMachinePool
+	if pool == nil || pool.Spec.NodeDrainTimeout == nil || pool.Spec.NodeDrainTimeout.Seconds() <= 0 {
+		return false
+	}
+
+	// if the draining succeeded condition does not exist
+	if conditions.Get(s.AzureMachinePoolMachine, clusterv1.DrainingSucceededCondition) == nil {
+		return false
+	}
+
+	now := time.Now()
+	firstTimeDrain := conditions.GetLastTransitionTime(s.AzureMachinePoolMachine, clusterv1.DrainingSucceededCondition)
```

Comment:
Clarification:`DrainSucceededCondition` is marked false if there is an error as well [here](https://github.com/devigned/cluster-api-provider-azure/blob/6dae9549dd2949ace7ba1384eaf7e86798edc884/azure/scope/machinepoolmachine.go#L319) with a different severity and message. Will this update the transition time?

Reply:
This is a great catch. Based on the following, I think you are correct:
- Check in transition to see if conditions match
  - https://github.com/kubernetes-sigs/cluster-api/blob/59c620c36aead9e76157210a063c05d33c3bd276/util/conditions/setter.go#L53-L57
- Check sameness
  - https://github.com/kubernetes-sigs/cluster-api/blob/59c620c36aead9e76157210a063c05d33c3bd276/util/conditions/setter.go#L167-L175

Reply:
Updated the `MarkFalse` error location with a hefty comment and a guard to ensure the condition is `nil` before setting.